### PR TITLE
[Core][C++ Worker]c++ worker support std::any param type for cross language

### PR DESCRIPTION
**Pull request by larrylian**
## Why are these changes needed?
the paramter and return type of C++ Worker support `std::any`  for cross language.
C++ Worker `std::any` is map to Java object.

TODO: The current pull request encounters errors when compiling on Windows, so Windows support is temporarily not available. I will work on adapting it to Windows and merge the Windows support once it is ready.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: cpp/include/ray/api/msgpack_adaptor.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <any>
+#include <msgpack.hpp>
+
+namespace msgpack {
+namespace adaptor {
```
Comment (SongGuyang) [inline cpp/include/ray/api/msgpack_adaptor.h:21]: Why we need this adaptor? Can you add some comments or links for this?

**Review by SongGuyang — COMMENTED** [review]
Overall looks good

File: cpp/src/ray/test/serialization_test.cc [hunk 2]
```diff
@@ -40,6 +69,37 @@ TEST(SerializationTest, TypeHybridTest) {
 
   EXPECT_EQ(in_arg1, out_arg1);
   EXPECT_EQ(in_arg2, out_arg2);
+
+  // Test std::any
+  msgpack::sbuffer buffer3 = ray::internal::Serializer::Serialize(std::make_tuple(
+      in_arg1, in_arg2, in_arg3, in_arg4, in_arg5, in_arg6, in_arg7, in_arg8));
+
+  std::vector<std::any> result =
```
[the remark this answers is not in the document]
Comment (larrylian) [reply cpp/src/ray/test/serialization_test.cc:77]: Yes, msgpack support tuple.   tuple -> std::vector<std::any>

File: java/test/src/main/java/io/ray/test/CrossLanguageInvocationTest.java [hunk 3]
```diff
@@ -268,6 +268,43 @@ public void testCallingCppActor() {
...
+          "Hello World!", // String
+          1.234f, // Float
+          1.234, // Double
+          "example binary".getBytes() // byte[]
+        };
+    ObjectRef<Object[]> ref_objs =
+        actor.task(CppActorMethod.of("EchoAnyArray", Object[].class), inputs).remote();
+    Object[] objs_result = ref_objs.get();
+    // Assert.assertEquals(objs_result, inputs);
+    Assert.assertEquals((Boolean) objs_result[0], inputs[0]);
```
[the remark this answers is not in the document]
Comment (larrylian) [reply java/test/src/main/java/io/ray/test/CrossLanguageInvocationTest.java:295]: Fixed.

File: cpp/include/ray/api/msgpack_adaptor.h (hunk 1 above)
Comment (larrylian) [reply cpp/include/ray/api/msgpack_adaptor.h:21]: Yes.   Added a comment.

File: cpp/src/ray/test/cluster/counter.h [hunk 4]
```diff
@@ -57,9 +58,47 @@ class Counter {
...
 
-  std::string echoString(const std::string &str) { return str; }
+  std::string EchoString(const std::string &str) { return str; }
+
+  std::vector<std::string> EchoStrings(const std::vector<std::string> &strings) {
+    return strings;
+  }
+
+  std::vector<std::any> EchoAnyArray(const std::vector<std::any> &anys) {
+    // check bool
```
[the remark this answers is not in the document]
Comment (larrylian) [reply cpp/src/ray/test/cluster/counter.h:70]: Fixed

**Review by SongGuyang — APPROVED** [review]
LGTM

File: cpp/include/ray/api/msgpack_adaptor.h [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    default:
+      throw msgpack::type_error();
+    }
+    return o;
+  }
+};
+
+}  // namespace adaptor
+}  // namespace msgpack
+#endif
```
Comment (SongGuyang) [inline cpp/include/ray/api/msgpack_adaptor.h:88]: Please add a empty line at the end.

**Review by SongGuyang — APPROVED** [review]

**Pull request merged**